lib/apps.fnl

```diff
@@ -0,0 +1,263 @@
+
+(local states
+       {:general-app {:enter-app  general->enter
+                      :leave-app  noop
+                      :launch-app ->launch
+                      :close-app  ->close}
+        :in-app      {:enter-app  in-app->enter
+                      :leave-app  in-app->leave
+                      :launch-app ->launch
+                      :close-app  ->close}})
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Watchers, Dispatchers, & Logging
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(local app-events
+       {hs.application.watcher.activated   :activated
+        hs.application.watcher.deactivated :deactivated
+        hs.application.watcher.hidden      :hidden
+        hs.application.watcher.launched    :launched
+        hs.application.watcher.launching   :launching
+        hs.application.watcher.terminated  :terminated
+        hs.application.watcher.unhidden    :unhidden})
+
+
+(fn watch-apps
+  [app-name event app]
+  (let [event-type (. app-events event)]
+    (if (= event-type :activated)
+        (enter app-name)
+        (= event-type :deactivated)
+        (leave app-name)
+        (= event-type :launched)
+        (launch app-name)
+        (= event-type :terminated)
+        (close app-name))))
+
+(fn active-app-name
+  []
+  (let [app (hs.application.frontmostApplication)]
+    (if app
+        (: app :name)
+        nil)))
+
+(fn start-logger
+  [fsm]
+  (atom.add-watch
+   fsm.state :log-state
+   (fn log-state
+     [state]
+     (log.df "app is now: %s" (and state.app state.app.key)))))
+
+(fn proxy-actions
+  [fsm]
+  (atom.add-watch fsm.state :actions
+                  (fn action-watcher
+                    [state]
+                    (emit state.action state.app))))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; API Methods
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(fn get-app
+  []
+  (when fsm
+    (let [state (atom.deref fsm.state)]
+      state.app)))
+
+(fn subscribe
+  [f]
+  (let [key (gen-key)]
+    (atom.add-watch actions key f)
+    (fn unsubscribe
+      []
+      (atom.remove-watch actions key))))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Initialization
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(fn init
+  [config]
+  (let [active-app (active-app-name)
+        initial-state {:apps config.apps
+                       :app nil
+                       :status :general-app
+                       :unbind-keys nil
+                       :action nil}
+        app-watcher (hs.application.watcher.new watch-apps)]
+    (set fsm (statemachine.new states initial-state :status))
+    (start-logger fsm)
+    (proxy-actions fsm)
+    (enter active-app)
+    (: app-watcher :start)
+    (fn cleanup []
+      (: app-watcher :stop))))
```

Comment:
not clear to me why it returns a function and how this suppose to work

Reply:
Discussed this a bit in a conversation but recapping here for posterity:

The main idea here is to have a function used to create a resource return a function to stop or destroy the resource. It's inspired by the RxJS Disposable and RxJS.Observable.create method. I like it because I don't have to deal with mutable resources and think about their state. I create a resource and get a function to shut it down. The returned function is safe so that I can call it multiple times without throwing errors. This is very helpful for managing a sub-menu's keybindings. Rather than having to manage a hotkey modal instance, I just get a function to remove the bindings. Whenever our state machine transitions to a close of the modal we can call that function. If per-chance multiple transitions are triggered consecutively, it doesn't matter how many times we call it, the resource will be destroyed and we don't have to manually change state or check state.

A code example:

```
(local kill-timer (set-interval tick-fn 200))
;; Anytime after...
(kill-timer)
(kill-timer)
```

Here set-interval returns a function to clear the interval. We don't need to manage the timer instance, in our code we only need to focus on when we wish to call kill-timer. 

We can call it multiple times like the example without throwing any errors or messing with any resources. If I returned the timer instance and called something like `(clear-interval my-interval)` multiple times it may throw an error due to the state of the resource changing between calls.